github/client.go

```diff
@@ -0,0 +1,196 @@
+	client.GraphQL = gh_graphql.NewClient(graphqlClient)
+
+	return client, nil
+}
+
+// clientRoundTripper returns a RoundTripper that logs requests and responses to the GitHub API.
+// You can pass a custom RoundTripper to use a different transport, or nil to use the default transport.
+func clientRoundTripper(clientType string, l zerolog.Logger, next http.RoundTripper) http.RoundTripper {
+	if next == nil {
+		next = http.DefaultTransport
+	}
+
+	return &loggingTransport{
+		transport:  next,
+		logger:     l,
+		clientType: clientType,
+	}
+}
+
+// loggingTransport is a RoundTripper that logs requests and responses to the GitHub API.
+type loggingTransport struct {
+	transport  http.RoundTripper
+	logger     zerolog.Logger
+	clientType string
+}
+
+// RoundTrip logs the request and response details.
+func (lt *loggingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+
+	l := lt.logger.With().
+		Str("client_type", lt.clientType).
+		Str("method", req.Method).
+		Str("request_url", req.URL.String()).
+		Str("user_agent", req.Header.Get("User-Agent")).
+		Logger()
+
+	resp, err := lt.transport.RoundTrip(req)
+	duration := time.Since(start)
+
+	l = l.With().
+		Int("status", resp.StatusCode).
+		Str("duration", duration.String()).
+		Logger()
+
+	if err != nil || resp.StatusCode != http.StatusOK {
+		// Probably a rate limit error, let the rate limit library handle it
+		return resp, err
+	}
+
+	// Process rate limit headers
+	callsRemainingStr := resp.Header.Get("X-RateLimit-Remaining")
+	if callsRemainingStr == "" {
+		callsRemainingStr = "0"
+	}
+	callLimitStr := resp.Header.Get("X-RateLimit-Limit")
+	if callLimitStr == "" {
+		callLimitStr = "0"
+	}
+	callsUsedStr := resp.Header.Get("X-RateLimit-Used")
+	if callsUsedStr == "" {
+		callsUsedStr = "0"
+	}
+	limitResetStr := resp.Header.Get("X-RateLimit-Reset")
+	if limitResetStr == "" {
+		limitResetStr = "0"
+	}
+	callsRemaining, err := strconv.Atoi(callsRemainingStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert callsRemaining header to int: %w", err)
+	}
+	callLimit, err := strconv.Atoi(callLimitStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert callLimit header to int: %w", err)
+	}
+	callsUsed, err := strconv.Atoi(callsUsedStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert callsUsed header to int: %w", err)
+	}
+	limitReset, err := strconv.Atoi(limitResetStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert limitReset header to int: %w", err)
+	}
+	limitResetTime := time.Unix(int64(limitReset), 0)
+
+	l = l.With().
+		Int("calls_remaining", callsRemaining).
+		Int("call_limit", callLimit).
+		Int("calls_used", callsUsed).
+		Time("limit_reset", limitResetTime).
+		Str("response_url", resp.Request.URL.String()).
+		Logger()
+
+	mockRequest := strings.TrimPrefix(req.Header.Get("Authorization"), "Bearer ") == MockToken
+
+	if !mockRequest && callsRemaining <= 10 && callsRemaining%2 == 0 {
```

Comment:
If this is the only time `mockRequest` is used, then it seems unnecessary to expose this functionality at all.

Reply:
Also just in general, if the mock request functionality is used. Would it make sense to expose it as a custom HTTP header included in the request rather than creating a side-effect on existing functionality?

`flakeguard-mock-request: true` seems more explicit than setting the bearer token's value to something specific.

Reply:
Agreed.